File: KSC2015_ProblemSet_n_Solution/Competition2015/04_KISTI/Code/Fortran/serial/poisson.f90
program main

      implicit none

      real(8), parameter :: PI = 3.1415926535897932384626
      integer(4) :: numgrid_x, numgrid_y
      integer(4) :: matrixDOF
      integer(4) :: count, ii, jj, kk, maxiteration
      real(8) :: length_x, length_y, gridsize, tolerance
      real(8), allocatable :: poissonmatrix(:), rhs(:), solution(:), coordinate(:)
      character(256) ::  myfilename
!FILE *fp;

      maxiteration = 20000
      tolerance = 1.0e-10
      gridsize = 0.01
      length_x = 1.0
      length_y = 1.0

      numgrid_x = nint(length_x/gridsize) + 1
      numgrid_y = nint(length_y/gridsize) + 1
      matrixDOF = numgrid_x * numgrid_y

      write(*,*) '[Poisson] Geometry and matrix size initialized.'

      allocate( coordinate(matrixDOF*2) ) 
      allocate( rhs(matrixDOF) )

      count = 1

      do ii = 1, numgrid_x
          do jj = 1, numgrid_y
              coordinate(2*count-1) = (ii-1)*gridsize;
              coordinate(2*count) = (jj-1)*gridsize;
              rhs(count) = sin(coordinate(2*count-1)/length_x*PI)*sin(coordinate(2*count)/length_y*PI)*gridsize*gridsize
              count = count+1
          enddo
      enddo

      write(*,*) '[Poisson] Geometry and rhs constructed.'

      allocate( poissonmatrix(matrixDOF*matrixDOF) )

      do ii=1, matrixDOF*matrixDOF
          poissonmatrix(ii) = 0.0
      enddo

      call construct_poissonmatrix(1, 1, numgrid_x, numgrid_x, numgrid_y, poissonmatrix)

      write(*,*) '[Poisson] Poisson matrix constructed.'

      allocate( solution(matrixDOF) )

      do ii=1, matrixDOF
          solution(ii) = 1.0;
      enddo

      write(*,*) '[Poisson] Start solving equations.'

      call cgsolver(matrixDOF, poissonmatrix, rhs, solution, maxiteration, tolerance)

      write(*,*) '[Poisson] Solution obtained.'

      call system('rm -rf result')
      call system('mkdir result')
      open(11, file='./result/solution.dat', form='formatted')
      
      count = 1
      do ii=1, numgrid_x
          do jj=1, numgrid_y
              write(11,101) coordinate(2*count-1), coordinate(2*count), solution(count)
              count = count + 1
          enddo
      enddo
101   format(2(f9.5,x),f12.6)

      close(11)

      write(*,*) '[Poisson] Solution printed.'

      deallocate(solution)
      deallocate(poissonmatrix)
      deallocate(rhs)
      deallocate(coordinate)

      write(*,*) '[Poisson] Memory deallocated.'

end program

File: KSC2015_ProblemSet_n_Solution/Competition2015/04_KISTI/Code/C/serial/poisson.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "matrixconstructor.h"
#include "cgsolver.h"

#define PI 3.141592653589793

int main(int argc, char **argv) {

	int numgrid_x, numgrid_y;
	int matrixDOF;  
	int count, ii, jj, kk, maxiteration;  
	double length_x, length_y, gridsize, tolerance;
	double *poissonmatrix, *rhs, *solution, *coordinate;
	char myfilename[256];
	FILE *fp;

	maxiteration = 20000;
	tolerance = 1.0e-10;
	gridsize = 0.01;
	length_x = 1.0;
	length_y = 1.0;

	numgrid_x = (int)round(length_x/gridsize) + 1;
	numgrid_y = (int)round(length_y/gridsize) + 1;
	matrixDOF = numgrid_x*numgrid_y;

	printf("[Poisson] Geometry and matrix size initialized.\n");

	coordinate = (double*)malloc(matrixDOF*2*sizeof(double));
	rhs = (double*)malloc(matrixDOF*sizeof(double));

	count = 0;

	for(ii=1; ii<=numgrid_x; ii++)
	{
		for(jj=1; jj<=numgrid_y; jj++)
		{
			coordinate[2*count] = (ii-1)*gridsize;
			coordinate[2*count+1] = (jj-1)*gridsize;
			rhs[count] = sin(coordinate[2*count]/length_x*PI)*sin(coordinate[2*count+1]/length_y*PI)*gridsize*gridsize;
			count++;
		} 
	}
	
	printf("[Poisson] Geometry and rhs constructed.\n");

	poissonmatrix = (double*) malloc(matrixDOF*matrixDOF*sizeof(double)); 

	for(ii=0; ii<matrixDOF*matrixDOF; ii++)	
		poissonmatrix[ii] = 0.0;

	construct_poissonmatrix(1, 1, numgrid_x, numgrid_x, numgrid_y, poissonmatrix);

	printf("[Poisson] Poisson matrix constructed.\n");

	solution = (double*) malloc(matrixDOF*sizeof(double));

	for(ii=0; ii<matrixDOF; ii++)
		solution[ii] = 1.0;

	printf("[Poisson] Start solving equations.\n");

	cgsolver(matrixDOF, poissonmatrix, rhs, solution, maxiteration, tolerance);

	printf("[Poisson] Solution obtained.\n");

	system("rm -rf result\n");
	system("mkdir result\n");
	sprintf(myfilename,"./result/solution.dat");
	fp = fopen(myfilename, "wt");
	
	count = 0;
	for(ii=1; ii<=numgrid_x; ii++)
	{
		for(jj=1; jj<=numgrid_y; jj++)
		{
			fprintf(fp,"%9.5f%10.5f%13.6f\n", coordinate[2*count], coordinate[2*count+1], solution[count]);
			count++;
		}
	}

	fclose(fp);

	printf("[Poisson] Solution printed.\n");

	free(solution);
	free(poissonmatrix);
	free(rhs);
	free(coordinate);

	printf("[Poisson] Memory deallocated.\n");
	return 0;
}


